README.md
Why another database for macroeconomic data?
============================================

- Machine-readable datafeeds for economic data are growing ([FRED](https://research.stlouisfed.org/docs/api/fred/), 
  [quandl](https://blog.quandl.com/api-for-economic-data), 
  [OECD](https://data.oecd.org/api), 
  [World Bank](https://datahelpdesk.worldbank.org/knowledgebase/topics/125589), 
  [EIA](https://www.eia.gov/opendata/)). Please take a minute to fill [our poll about economics datasources](https://goo.gl/2wY43R) and see the results.  

- However, some data is still left in the dark. Russian macroeconomic statistics seems very fragmented (HTML, Word, Excel are common dessimination formats). This is a roadblock to reproducible analysis as dirty data escalates costs of model maintenance.      

- ```mini-kep``` aims to remove this roadblock by providing 
  [public API for Russian macroeconomic data](http://mini-kep.herokuapp.com/) 
  and examples of economic research/business planning/marketing  problems solved in python pandas or R.
  
 
User case
=========

We assume an end user has some experience with data from FRED or quandl.

The user wants:

- a clean dataset with latest data from different sources
- browse what data is available
- read this data on a local machine:
   - as pd.DataFrame 
   - as R dataframe  
- quickly draw some charts like one below: 

[![](http://datachart.cc/images/rub_oil.png)](http://datachart.cc/)

More on user case [here](specification/usercase.md).

Read next
=========

- [Specification pages](https://github.com/mini-kep/intro/blob/master/specification/README.md)
- [Development notes](DEV.md) 
- [Changelog](changelog.md)
- [New user checklist](https://github.com/mini-kep/intro/wiki/New-user-checklist)



CHANGELOG.md
Changelog
=========

- **2017-10-06** db POST/GET implemented <https://minikep-db.herokuapp.com/api/datapoints?name=BRENT&freq=d>
- **2017-09-28** new data pipeline description [open for comments](https://github.com/mini-kep/intro/issues/14) 
- **2017-09-28** <https://github.com/mini-kep/parsers> provide data from three parsers (macro, oil, fx)
- **2017-09-07** flask app <http://mini-kep.herokuapp.com/> relays parsing results from 
                 <https://github.com/mini-kep/parser-rosstat-kep>                  


specification/include/database_tests.py
"""Specification tests for deployed application. 

These tests are a quick check for "Does appliation generally fit the 
<spec_database.py> specification?"

These tests call http address of deployed apllication. They may be used to check
both flask and django application.
   
These tests DO NOT substitute internal unittests inside the application. 
The intenal tests should not call http, but rather use app.test_client()
in flask or similar command in django.

"""
import json

import requests

HOST = 'http://127.0.0.1:5000'
POST_URI = HOST + '/api/incoming'
BASE_URI = HOST + '/api/datapoints'

DATA = json.dumps([{'date': '2017-09-18', 'freq': 'd', 'name': 'BRENT', 'value': 55.5},
                   {'date': '2017-09-15', 'freq': 'd', 'name': 'BRENT', 'value': 56.18},
                   {'date': '2017-09-14', 'freq': 'd', 'name': 'BRENT', 'value': 56.76},
                   {'date': '2017-09-13', 'freq': 'd', 'name': 'BRENT', 'value': 55.52},
                   {'date': '2017-09-12', 'freq': 'd', 'name': 'BRENT', 'value': 55.06},
                   {'date': '2017-09-11', 'freq': 'd', 'name': 'BRENT', 'value': 54.2}]
)


class TestPOST():
    # success
    def test_posting_valid_data(self):
        """Return an empty JSON on success"""
        data = DATA
        headers = {'content-type': 'application/json'}
        response = requests.post(POST_URI, json=data, headers=headers)
        assert response.json() == []

    # fail
    def test_posting_nothing(self):
        """Results in a 400 error"""
        response = requests.post(POST_URI)
        # FIXME:'123qwe'
        assert response.status_code == '123qwe'



class TestGET():
    # happy paths
    def test_request_response_with_valid_params(self):
        """Completed successfully with OK status code"""
        payload = {'name': 'BRENT', 'freq': 'd'}
        response = requests.get(BASE_URI, params=payload)
        assert response.ok

    def test_getting_BRENT(self):
        """Return a list with items that have the requested variable"""
        payload = {
            'name': 'BRENT',
            'freq': 'd'
        }
        response = requests.get(BASE_URI, params=payload)
        assert response.json()[0]['name'] == 'BRENT'

    # sad paths
    def test_request_with_empty_name(self):
        """Return 400 error with empty name parameter"""
        payload = {'name': '', 'freq': 'm'}
        response = requests.get(BASE_URI, params=payload)
        assert response.status_code == 400

    def test_request_with_empty_frequency(self):
        """Return 400 error with empty freq parameter"""
        payload = {'name': 'BRENT', 'freq': ''}
        response = requests.get(BASE_URI, params=payload)
        assert response.status_code == 400

    def test_getting_data_not_found(self):
        """Return response with empty json"""
        payload = {'name': 'FOOBAR', 'freq': 'm'}
        response = requests.get(BASE_URI, params=payload)
        assert response.json() == []

markdown/md_helper.py
"""Helper to get Travis/Codecov badges in markdown table."""


class Markdown:
    def _link(what, link):
        return f'[{what}]({link})'
    
    def _img(img_url, alt=''):
        return f'![{alt}]({img_url})'
    
    def linked_image(img_url, link_url):
        img_str = Markdown._img(img_url)
        return Markdown._link(what=img_str, link=link_url)        

class App:
    def badge(self):
        return Markdown.linked_image(self.svg, self.link)


class Travis(App):  
    def __init__(self, repo_name):
        self.link = f'https://travis-ci.org/{repo_name}'
        self.svg = f'{self.link}.svg?branch=master'

assert Travis('mini-kep/parser-rosstat-kep').badge() == ("[![]"
    "(https://travis-ci.org/mini-kep/parser-rosstat-kep.svg?branch=master)]"
    "(https://travis-ci.org/mini-kep/parser-rosstat-kep)")

           
class Codecov(App):    
    def __init__(self, repo_name):
        self.link = f'https://codecov.io/gh/{repo_name}'
        self.svg = f'{self.link}/branch/master/graphs/badge.svg'

        
class Repo:    
    def __init__(self, repo_name):
        self.name = repo_name
        self.name_in_org = f'mini-kep/{repo_name}' 
        self.url = f'https://github.com/mini-kep/{self.name}'
    
    def badge_codecov(self):
        return Codecov(self.name_in_org).badge()
    
    def badge_travis(self):
        return Travis(self.name_in_org).badge()

    def link(self):
        return Markdown._link(self.name, self.url)


def to_markdown(table):
    """Translate *table* list of strings to markdown. 
       Use first row in *table* as header.       
    """    
    header = table[0]
    horiz = ["-" * len(x) for x in header]
    body = table[1:]
    table = [header, horiz]
    table.extend(body)
    
    def add_dividers(row):
        inner_part = " | ".join(row) 
        return f"| {inner_part} |"
    
    table = [add_dividers(row) for row in table]    
    return '\n'.join(table)

def row_with_badges(repo_name):
    repo = Repo(repo_name)
    return [repo.link(), 
            repo.badge_travis(), 
            repo.badge_codecov()] 
           
def row_with_link(repo_name):
    repo = Repo(repo_name)
    return [repo.link(), '', '']
   

if __name__ == "__main__":
    repo_names = ['parser-rosstat-kep', 
                  'parsers',
                  'db',
                  'frontend-app',
                  'full-app']
    table = [['Repo', 'Tests', 'Coverage']]
    table.extend([row_with_badges('parser-rosstat-kep')])
    table.extend([row_with_badges('parsers')])
    table.extend([row_with_link('db')])
    table.extend([row_with_badges('frontend-app')])
    table.extend([row_with_badges('full-app')])
    md_table = to_markdown(table)
    print(md_table)


pipeline/older_versions/pipeline_demo.py
# -*- coding: utf-8 -*-

import json
from pprint import pprint
import time


import pandas as pd

d = {'RUR_EUR_eop': pd.Series([48.1, 49.35, 49.05, 49.51, 47.27, 45.83, 47.9, 48.63, 49.95, 54.64, 61.41, 68.34, 78.11, 68.69, 63.37, 56.81, 58.01, 61.52, 64.65, 75.05, 74.58, 70.75, 70.39, 79.7]), 
     'CPI_rog': pd.Series([100.6, 100.7, 101.0, 100.9, 100.9, 100.6, 100.5, 100.2, 100.7, 100.8, 101.3, 102.6, 103.9, 102.2, 101.2, 100.5, 100.4, 100.2, 100.8, 100.4, 100.6, 100.7, 100.8, 100.8])}
ix = pd.date_range(start='2014-01-01', periods=12*2, freq='M')


# 0. Here is something a user wants, Jane is working with dfm DataFrame  

dfm = pd.DataFrame(d).set_index(ix)

# 1. parser jobs

# 1.1 parser1 parsed something and is ready to provide the data

to_serialise = json.loads(dfm.to_json(None,'index', date_format='iso'))

observations = [dict(date=k.replace('T00:00:00.000Z',''),
                     freq='m',
                     varname = a,
                     value = b) 
                for k, v in to_serialise.items() 
                for a, b in v.items()]

parcer_output1 = json.dumps(observations, indent=4)
print()
print("This goes from parser 1 to database")
print("====================================")
print(parcer_output1)

# 1.2 parser2 parsed something else and is ready to provide the data too
#     it is likely to be different variables, but I'll show same var at different frequency

observations = [
    {"date": "2014-03-31", "freq": "q", 
     "varname": "CPI_rog", "value": 102.3}
    ]

parcer_output2 = json.dumps(observations, indent=4)

print()
print("This goes from parser 2 to database")
print("===================================")
print(parcer_output2)



# 2. the database accepted 1.1 and 1.2 this and stores the data in whichever 
#    is most appropriate format 

# ...

# 3. the user sends a query to the database

user_query = dict(varnames=['CPI_rog', 'RUR_EUR_eop'],
                  freq='m',
                  start='2014-01', # optional
                  end='2015-12' # optional
                  )

print()
print("User query")
print("==========")
pprint(user_query)



# 3. based on a query, the database will send this json to the user

# we use dfm.to_json() to show, but the real json will be constructed from the database  
# dfm will not be available at this time
json_query_result = dfm.to_json() 
print()
print("App response to query: json with epoch timestamps")
print("=================================================")
print("This format is default input \nto user reader function pd.read_json()")
pprint(json.loads(json_query_result))


#4. the user accepts json and renders it locally to dataframe type  
dfm_at_user = pd.read_json(json_query_result)
print()
print("User's local dataframe")
print("======================")
print(dfm_at_user)
flag = dfm_at_user.equals(dfm) 
assert dfm_at_user.equals(dfm) 


print("Identical to source data:", flag)


# concerns:

# 1. when quering for parser result must decide which dates are newest
#    must not ask for all of data from long ago
    
   
# 2. namespaces can be useful, eg ru/CPI_rog, us/CPI_rog, eu/CPI_rog    
#    can map parser names to namespaces


# notes:
    
# even with one time series, read_jsom retrurns Dataframe, not Series
f = """{"CPI_rog": {"1391126400000": 100.6, 
             "1393545600000": 100.7, 
             "1396224000000": 101.0,
             }}"""
assert isinstance(pd.read_json(f), pd.DataFrame)

# converting epoch
epoch = "1451520000000"
assert "2015-12-31" == time.strftime('%Y-%m-%d', 
                       time.localtime(int(epoch) / 10 ** 3))


# can use this learn about 'orient' paramater 
#for orient in ('split','records','index','columns','values'):
#    print()
#    this_json = dfm.to_json(None, orient)
#    size = len(str(this_json))
#    print(orient, size)
#    print(this_json)


pipeline/pipeline.py
"""Fetch macroeconomic data from external sources and provide it to user 
   in a clean, friendly way.

This document mimics the data pipline from parsers to end-user API:
    
   1. parsers make data from different sources available
   2. database stores the data
   3. standard (REST) API returns query results for full-syntax query
   4. custom end-user API returns json readable by pd.read_json(),
      the query syntax is simplified for economists to understand
      
You can also see it from the user end:
   - the user wants an economic time series     
   - she queries the api like 
   
   df = pd.read_json('http://mini-kep.herokuapp.com/ru/series/CPI/m/rog/2015')
   
   # that is deliever time series for Russian (ru) consumer price index (CPI)
   # at monthly frequency (m) measured as rate of growth to previous period 
   # (rog) starting from year (2015)
   
   - the query goes to flask app that decodes this into more standard 
   select operation (a GET in REST API call to the database)
   
   - the database should contain such data
   
   - in the database the data was loaded from the parsers which read 
     the original data sources (statuc files and other APIs). 
   
   
   Questions, comments, concerns highly appreciated.
   
   So far:
       > what is implemented and what is not? - added comments along the lines, 
         but we more or less have the parsers and end-user API, no database 
         layer yet (there are prototypes but they are not workable yet)
       > can we live without REST API around the database (#3) - yes we can, 
         it is optional, but I think it is good abstration that can help build 
         other services around the db
       > can you go without jargon like REST and 'reddit'? - I tried and 
         edited ;)    
          
"""


import json
import pandas as pd

# 1. The parsers are ready to provide data
# this dataset is generated by Dataset.yield_dicts() 
# at <https://github.com/mini-kep/parsers/blob/master/parsers/runner.py>

with open('dataset.json') as f:
    data = json.load(f)

# available variable names 
names = set([d['name'] for d in data])


# 2. the database accepted and stores the data 
#    in whichever is most appropriate format 

# what we have have for the database:
    
#    here we tried CRUD methods in pure SQL Alchemy:
#    <https://github.com/mini-kep/db/blob/master/clientdb.py>
#    that kind of hanged here:
#    <https://github.com/mini-kep/db/issues/3>

#    also there is Django app with a simple db schema inside
#    <https://github.com/mini-kep/full-app>
#    but it all stopped here:
#    <https://github.com/mini-kep/full-app/issues/28>
#    if you knwo Djngo well, youcan probablu speed up the process

# code below mimicks a 'select' operation performed by Django ORM or SQLAlchemy

def is_wanted(datapoint, user_query):
    """Does the datapoint comply with user query?"""
    positive_conditions = []
    positive_conditions.append(datapoint['name']==user_query['name'])
    positive_conditions.append(datapoint['freq']==user_query['freq'])
    positive_conditions.append(datapoint['date']>=user_query['start'])
    #skipping this to simplify
    #positive_conditions.append(datapoint['date']<=user_query['end'])
    return all(positive_conditions)


def select(user_query_rest):
    # using global *data*
    return [d for d in data if is_wanted(d, user_query_rest)]


# 3. 'Rest' API for the database
#    The database should accept/send data in json for read/write operations 

# the user sends a query to the database
user_query_rest_sample = dict(name='CPI_rog',
                  freq='m',
                  start='2014-01-01', 
                  end='2015-12-31')

# the server replies
rest_reply = select(user_query_rest_sample)


# 4. Custom API 
# - uses custom, 'reddit-like' notation for end-user API call, as in 
#   <http://mini-kep.herokuapp.com/ru/series/CPI/m/rog/2015>
#    flask knows to decipher long URL and retruns a dict with parameters, try it!
#    the next thing flask should do is return real data 
#
# - returns json readable by pd.read_json()  

def to_json(dicts):    
    df = pd.DataFrame(dicts)
    df.date = df.date.apply(pd.to_datetime)
    df = df.pivot(index='date', values='value', columns='name')
    return df.to_json()

def to_string(pd_date):
    return pd_date.date().strftime('%Y-%m-%d')

def interpret_reddit(user_query_custom):
    name, freq, unit, start_year = user_query_custom.split("/")
    return dict(name=f'{name}_{unit}', 
                freq=freq, 
                start=to_string(pd.to_datetime(start_year)),
                end=to_string(pd.to_datetime(0).today())
                )
    
def custom_select(user_query_custom):
    # break down the *user_query_custom* to *user_query_rest*
    user_query_rest = interpret_reddit(user_query_custom)
    # query rest API
    return select(user_query_rest)

def server_response_custom(user_query_custom):
    return to_json(dicts=custom_select(user_query_custom))

if __name__ == "__main__":
    # this part is user scenario, getting some data inside dataframe    
    
    # the user sends a query to the database
    user_query_custom_sample = 'CPI/m/rog/2017'
    
    # the server replies
    user_result = server_response_custom(user_query_custom_sample)
    
    # the user recieves data and makes a dataframe
    user_df = pd.read_json(user_result)
    
    assert user_df.__str__() == \
"""            CPI_rog
2017-01-31    100.6
2017-02-28    100.2
2017-03-31    100.1
2017-04-30    100.3
2017-05-31    100.4
2017-06-30    100.6
2017-07-31    100.1"""
